.github/workflows/ci.yml

```diff
@@ -2,20 +2,169 @@ name: CI
 
 on:
   push:
+    branches:
+      - master
   pull_request:
+  release:
+    types: [published]
+
+env:
+  CARGO_TERM_COLOR: always
+  RUSTFLAGS: -D warnings
+  REGISTRY: ghcr.io
+  RUST_VERSION: 1.82.0
+  NIGHTLY_RUST_VERSION: nightly-2024-10-28
 
 jobs:
-  test:
-    name: cargo test
+  cancel-previous-runs:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Cancel Previous Runs
+        uses: styfle/cancel-workflow-action@0.9.1
+        with:
+          access_token: ${{ github.token }}
+
+  publish-codecov:
+    name: Check code coverage (branch)
     runs-on: ubuntu-latest
+    permissions: # Write access to push changes to pages
+      contents: write
+      pull-requests: write
     steps:
       - uses: actions/checkout@v4
-      - uses: dtolnay/rust-toolchain@stable
-        with:
-          components: clippy, rustfmt
-      - name: Cargo Format
-        run: cargo fmt --all --check
-      - name: Cargo Clippy
-        run: cargo clippy --workspace --all-targets --all-features -- -D warnings
-      - name: Cargo Test
-        run: cargo test --workspace --all-targets --all-features
\ No newline at end of file
+      - name: Install latest Rust
+        uses: dtolnay/rust-toolchain@master
+        with:
+          toolchain: ${{ env.NIGHTLY_RUST_VERSION }}
+
+      - name: Install cargo-llvm-codecov
+        uses: taiki-e/install-action@cargo-llvm-cov
+
+      - name: Code coverage report
+        run: cargo +${{ env.NIGHTLY_RUST_VERSION }} llvm-cov --all-features --lcov --branch --output-path lcov.info
+
+      - name: Setup LCOV
+        uses: hrishikesh-kadam/setup-lcov@v1
+
+      - name: Report code coverage
+        uses: zgosalvez/github-actions-report-lcov@v4
+        with:
+          coverage-files: lcov.info
+          minimum-coverage: 0 # for now we are not enforcing any minimum coverage.
+          artifact-name: code-coverage-report
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+          update-comment: true
+
+  lint-toml-files:
+    needs: cancel-previous-runs
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```

Comment:
we use `- uses: actions/checkout@v4` above and `- uses: actions/checkout@v2` here and in a few places below. Can we standardise on using v4?

Reply:
Oh that is a good catch I copied this from wallet, I will fix it there too :)